closes #22923 ruby files and serialization add missing knowledge check and links (#23114)
TheOdinProject/curriculum: ruby_programming/files_and_serialization/lesson_serialization.md

* [Zetcode's section on Input/Output in Ruby](http://zetcode.com/lang/rubytutorial/io/) should be another useful perspective on the material.
* [Ruby Monk's section on Serializing](https://web.archive.org/web/20160505174806/http://rubymonk.com/learning/books/4-ruby-primer-ascent/chapters/45-more-classes/lessons/104-serializing)
* [Short Example of Serialization](http://rubylearning.com/satishtalim/object_serialization.html) from Ruby Learning

### Knowledge Check

  * <a class='knowledge-check-link' href='https://thoughtbot.com/blog/io-in-ruby'>How would you go through the lines of a file using the IO object in the command line?</a>

>   * <a class='knowledge-check-link' href='https://www.sitepoint.com/choosing-right-serialization-format/'>How would you take the Array consisting of [1,2,3] and serialize the Array into a JSON string format?</a>

Reviewer: ```suggestion
  * <a class='knowledge-check-link' href='https://www.sitepoint.com/choosing-right-serialization-format/'>How would you take the Array `[1,2,3]` and serialize it into a JSON string format?</a>
```

  * <a class='knowledge-check-link' href='https://www.sitepoint.com/choosing-right-serialization-format/'>What are the benefits of YAML vs JSON vs MessagePack serialization formats?</a>
  * <a class='knowledge-check-link' href='http://ruby.bastardsbook.com/chapters/io/'>How would you check if a file exists using the File object in the command line?</a>

>   * <a class='knowledge-check-link' href='http://ruby.bastardsbook.com/chapters/io/'>In a Download directory containing 5 files. How would you count the number of files in that directory using the Dir object?</a>

Reviewer: ```suggestion
  * <a class='knowledge-check-link' href='http://ruby.bastardsbook.com/chapters/io/'>In a Download directory containing 5 files, how would you count the number of files in that directory using the Dir object?</a>
```

> 

Reviewer: Remove excess newlines

* [Zetcode's section on Input/Output in Ruby](http://zetcode.com/lang/rubytutorial/io/) should be another useful perspective on the material.
* [Ruby Monk's section on Serializing](https://web.archive.org/web/20160505174806/http://rubymonk.com/learning/books/4-ruby-primer-ascent/chapters/45-more-classes/lessons/104-serializing)
* [Short Example of Serialization](http://rubylearning.com/satishtalim/object_serialization.html) from Ruby Learning

### Knowledge Check

  * <a class='knowledge-check-link' href='https://thoughtbot.com/blog/io-in-ruby'>What does the IO object class do?</a>

>   * <a class='knowledge-check-link' href='https://www.youtube.com/watch?v=uS37TujnLRw'>How are data structures related to serialization?</a>

Reviewer: Can you explain this question to me and give an example of what kind of answer you're looking for? I'm not sure I understand it.

Author: For the first question in that point of the curriculum the goal of the question is check to see if they know what the IO object is supposed to do. In the linked article it states that the IO object wraps around input and output stream. It didn't strike me that it is necessary to remember what the STDIN, STDOUT, and STDERR constants point to. Rather just what it does as a object, although i can definitely make it more specific if need be. 

For the second question in the video he goes over different data structures such as an array or more complicated structures and how they are serialized. The question is to have them understand that data structures are related to serialization in that serialization is the method of storing/sending said data structure. 
My reasoning for these questions was that they ask about the core concepts I could make them more specific in terms of the scope of the question. Would it be better to phrase the second one along the lines of "How would i serialize a data structure such as an array according to the video?" and the first one something more specific such as "How would i go through a files lines using the IO object in the command line?". 
@wise-king-sullyman

Reviewer: Yes I think something along the lines of your suggested revision for the first question would be good, only big thing I would change as far as voice goes would be to speak in the second person rather than the first.

For the second question, I think you're on a better track with your suggestion but still not quite where I think is best as we don't want to frame Knowledge Checks in the context of / referencing specific resources.

If you'd like I can try and come up with a suggestion, but I don't have the mental energy to do so tonight and will have to circle back in the next few days.

Author: Let me chew on this for a few days and then get back to you with a better example of a question first. I think I have a better idea of what to do now.

Author: @wise-king-sullyman  Here are my updated questions for the knowledge check. Incorporated the feedback, rechecked them against the [wiki](https://github.com/TheOdinProject/curriculum/wiki/Learning-Outcomes-vs-Knowledge-Checks). Decided that the questions I had could be much more detailed and specified. Do these revised questions fit with the guidelines more so than the previous ones? 
1. How would you go through the lines of a file using the IO object in the command line?
2. How would you take the Array consisting of [1,2,3] and serialize the Array into a JSON string format?
3. What are the benefits of YAML vs JSON vs MessagePack serialization formats?
4. How would you check if a file exists using the File object in the command line?
5. In a Download directory containing 5 files. How would you count the number of files in that directory using the Dir object?

Reviewer: @linkonsat yes, those are great!

Author: @wise-king-sullyman Awsome! I'll resolve the merge conflict and update the PR with these questions tomorrow since it is a tad bit late.

Author: @wise-king-sullyman Conflict has been resolved and questions have been updated.

Reviewer: @linkonsat in the future once you've addressed all changes requested by a maintainer hit the "request re-review" button (screenshotted below in case you're not familiar) please, doing so helps us keep our fingers on the pulse of what all needs attention, and what each item in our notifications will be about.

![image](https://user-images.githubusercontent.com/7115053/145726221-d2500ecc-47ad-4482-b239-308ab4a6c6b9.png)

  * <a class='knowledge-check-link' href='https://www.sitepoint.com/choosing-right-serialization-format/'>What are three different serialization formats?</a>

>   * <a class='knowledge-check-link' href='https://www.sitepoint.com/choosing-right-serialization-format/'>What are the benefits of each type of serialization mentioned?</a>

Reviewer: I would like to see this question rewritten to not depend don't the context of the previous question. I.e. something along the lines of "benefits of YAML vs JSON vs MessagePack"

Author: Rewrote it to make it not reliant on the previous question. I thought it would be better to include serializations format at the end to make it clear what it's referring to without any context from the previous question.

Reviewer: In some cases it can be more about the implementation of a concept rather than a concept itself, but because that isn't always the case that phrasing wouldn't fit in something general purpose like the wiki.

>   * <a class='knowledge-check-link' href='http://ruby.bastardsbook.com/chapters/io/'>How would i use a File object to go through the lines in a file?</a>

Reviewer: ```suggestion
  * <a class='knowledge-check-link' href='http://ruby.bastardsbook.com/chapters/io/'>How can a File object be used to go through the lines of a file?</a>
```
